modules/mod-deep-learning/DeepModelManager.cpp

```diff
@@ -0,0 +1,554 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepModelManager.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "DeepModelManager.h"
+
+#include "CodeConversions.h"
+#include "NetworkManager.h"
+#include "IResponse.h"
+#include "Request.h"
+
+#include <wx/tokenzr.h>
+#include <wx/log.h>
+
+using namespace audacity;
+ 
+DeepModelManager::DeepModelManager() :
+   mAPIEndpoint("https://huggingface.co/api/")
+{
+   const std::string schemaPath = audacity::ToUTF8(
+      wxFileName(DeepModel::BuiltInModulesDir(), wxT("modelcard-schema.json"))
+                  .GetFullPath()
+   );
+   mModelCardSchema = parsers::ParseFile(schemaPath);
+}
+
+DeepModelManager& DeepModelManager::Get()
+{
+   static DeepModelManager manager;
+   return manager;
+}
+
+FilePath DeepModelManager::GetRepoDir(ModelCardHolder card) const
+{
+   static const std::string sep = "_";
+
+   FilePath repoDir = FileNames::MkDir( 
+      wxFileName(DeepModel::DLModelsDir(),
+                 card->author() + sep + card->name() 
+      ).GetFullPath()
+   );
+
+   return repoDir;
+}
+
+DeepModelHolder DeepModelManager::GetModel(ModelCardHolder card) const
+{
+   if (!IsInstalled(card))
+      throw ModelManagerException(XO("Model is not loaded."));
+
+   DeepModelHolder model = std::make_shared<DeepModel>();
+   model->SetCard(card);
+
+   // GetRepoDir won't work if the card is empty
+   wxFileName path = wxFileName(GetRepoDir(card), "model.pt");
+
+   // finally, load
+   model->Load(audacity::ToUTF8(path.GetFullPath()));
+
+   return model;
+}
+
+std::string DeepModelManager::GetMoreInfoURL(ModelCardHolder card) const
+{
+   return "https://huggingface.co/" + card->GetRepoID();
+}
+
+bool DeepModelManager::IsInstalled(ModelCardHolder card) const
+{
+   FilePath repoDir = wxFileName(card->local_path()).GetFullPath();
+
+   wxFileName modelPath = wxFileName(repoDir, "model.pt");
+   wxFileName metadataPath = wxFileName(repoDir, "metadata.json");
+
+   return (modelPath.FileExists() && metadataPath.FileExists());
+}
+
+void DeepModelManager::Install(ModelCardHolder card, ProgressCallback onProgress, CompletionHandler onCompleted)
+{ 
+   if (IsInstalled(card))
+      return ;
+
+   ProgressCallback progressHandler(
+   [this, card, handler {std::move(onProgress)}](int64_t current, int64_t expected)
+   {
+      // if the install has been cancelled, bail
+      // if we don't bail early, then calling the handler will 
+      // segfault.
+      if (!this->IsInstalling(card))
+         return;
```

Comment:
Now it's time to speak of the difficult topic of thread safety, where I see problems.

1. I understand now that this and other progress and completion callbacks ultimate are given to NetworkManager, which will execute them on non-main threads.
2. I see many uses of `CallAfter` in ModelCardPanel.cpp and ModelManagerPanel.cpp that had no explanation in the comments, but they make sense now that I understand this fact about threads.  An inner lambda, given to `CallAfter` by the outer lambda, means the inner lambda is sure to be enqueued to execute in the main thread later at idle time of the event loop, and that serializes those inner lambdas so there are no data races between them.
3. You may not have known this, but:  you don't need a window to invoke CallAfter.  You can also invoke this method on the global object, `wxTheApp` defined in `<wx/app.h>`.
4. In fact, I recommend you do exactly this, here at the low level of DeepModelManager.cpp, in all places that some callback is passed to NetworkManager.  Wrap all callbacks passed in from above in lambdas that pass them to `wxTheApp->CallAfter`; pass the wrapped callbacks, only, to NetworkManager; and then in DeepModelManager.h, where the callback `std::function` types are declared, comment that they are guaranteed to execute in the main thread.  Then, the callbacks upstairs in the Dialog files can be simplified.
5. Lambdas like this one should also be wrapped (DownloadModel in this case would wrap them, but they would remain unchanged here).  Because, as it is now, there is a dangerous data race.  `IsInstalling` uses `mResponseMap`, here in the thread pool, but the same map might be mutated concurrently in the main thread by `CancelInstall`.  Dangerous!  You could guard accesses with a mutex, but you can also just use CallAfter to serialize all the work in the main thread.

Reply:
I have more to say, about a better way to handle the lifetime of window objects.
1. The code comment above suggest to me you discovered a crash, and worked around it.  The workaround was not perfect, becase `this->IsInstalling(card)` might have transitioned from true above to false below, because of a thread context switch at just the wrong time.  Serializing all on the main thread will solve that.  Still, the need for the workaround points to something not right elsewhere in the program.
2. The only lambda that is ever wrapped inside this one is `onProgress` defined in `ModelCardPanel::OnInstall`, and I think you discovered a dangling pointer crash in that lambda because the `this` that was captured, no longer existed when the callback happened.  Am I right?
3. A solution for dangling pointers is to use a weak pointer.  I do not mean `std::weak_ptr` in this case, becuase it is not recommended to use standard smart pointers with certain wxWidgets objects, that have their own special memory management.  Instead, see this: https://docs.wxwidgets.org/3.0/classwx_weak_ref_3_01_t_01_4.html
4. So, better that `ModelCardPanel::OnInstall` should capture a `wxWeakRef` to `this`, and not `this` itself, in the lambda.  Then that allows a null pointer check inside that lambda, that the window still exists.  (No danger in that case that the non-null pointer will become stale because of the action of other threads.  All deletion of windows is done in the main thread.)
5. Really it is the responsibility of `onProgress`, defined there, to guard against capturing a dangling pointer.  Fix it there, then simplify here, removing this intermediate level of wrapping.
6. Should you do the same in the five other places that now use CallAfter in the *Dialog.cpp files?  I haven't proved yet to myself that it's necessary in all cases, but it is easy to do to enhance safety.  So, do that!